connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java

```diff
@@ -139,7 +141,8 @@ public void setup() throws Exception {
     public void close() {
         // stop all Connect, Kafka and Zk threads.
         connect.stop();
-        Block.resetBlockLatch();
+        // unblock everything so that we don't leak threads after each test run
+        Block.reset();
```

Comment:
It may be valuable to ensure that workers can shut down gracefully under these circumstances. Thoughts?

Reply:
I think that should take place in a test then, not in the cleanup.

The reason I bring this up is that if I were to assert that the clients/threads are all stopped immediately after Block.reset() (as implemented in #14783) there's no synchronization to ensure that cleanup takes place before the assertion fires. The "asynchronous cleanup" initiated by Block.reset could exceed the lifetime of the test, still leaking the threads but only temporarily.

Reply:
> I think that should take place in a test then, not in the cleanup.

It's easier to handle in a single method rather than copy over to 11 test cases, though. And I also don't necessarily see why `@After`-annotated classes need to be used exclusively for cleanup.

The concern about threads leaking (even if for a short period) beyond the scope of the test definitely seems valid. I've pushed a tweak that adds logic to wait for the blocked threads to complete in `Block::reset`. LMK if this seems clean enough; if not, I can bite the bullet and reverse the order of operations in `BlockingConnectorTest::close` and then see about adding more explicit checks for graceful worker shutdown in other places.

Reply:
> It's easier to handle in a single method rather than copy over to 11 test cases

Oh I see, I thought it would be sufficient to add one test case that called stop() to verify that one type of blocked thread still allows shutdown to complete, rather than verifying it for all of the different ways of blocking threads. That would have less coverage that the current tests on trunk.

> I've pushed a tweak that adds logic to wait for the blocked threads to complete in Block::reset.

I think this is probably the better solution. The leak tester can separately verify that resources were closed properly now that the test ensures the threads stop. :+1: